apps/dotcom/src/components/DocumentName/DocumentName.tsx

```diff
@@ -0,0 +1,295 @@
+
+	return (
+		<DocumentTopZoneContainer>
+			{isDocumentNameVisible && <DocumentNameInner />}
+			{isOffline && <OfflineIndicator />}
+		</DocumentTopZoneContainer>
+	)
+})
+
+export const DocumentNameInner = track(function DocumentNameInner() {
+	const [state, setState] = useState<NameState>({ name: null, isEditing: false })
+	const actions = useActions()
+	const forkAction = actions[FORK_PROJECT_ACTION]
+	const saveFileAction = actions[SAVE_FILE_COPY_ACTION]
+	const editor = useEditor()
+	const msg = useTranslation()
+
+	return (
+		<div className="tlui-document-name__inner">
+			<DocumentNameEditor state={state} setState={setState} />
+			<DropdownMenu.Root id="document name">
+				<DropdownMenu.Trigger>
+					<Button
+						type="icon"
+						icon="chevron-down"
+						className="tlui-document-name__menu tlui-menu__trigger flex-none"
+					/>
+				</DropdownMenu.Trigger>
+				<DropdownMenu.Content align="end" alignOffset={4} sideOffset={6}>
+					<DropdownMenu.Group>
+						<DropdownMenu.Item
+							type="menu"
+							id="rename"
+							onClick={() => setState((prev) => ({ ...prev, isEditing: true }))}
+						>
+							<span className={'tlui-button__label' as any}>{msg('action.rename')}</span>
+						</DropdownMenu.Item>
+						<DropdownMenu.Item
+							type="menu"
+							id="copy-link"
+							onClick={() => {
+								const shareLink = getShareUrl(
+									window.location.href,
+									editor.getInstanceState().isReadonly
+								)
+								navigator.clipboard.writeText(shareLink)
+							}}
+						>
+							<span className={'tlui-button__label' as any}>Copy link</span>
+						</DropdownMenu.Item>
+						<DropdownMenu.Item
+							type="menu"
+							id="save-file-copy"
+							onClick={() => saveFileAction.onSelect('document-name')}
+						>
+							<span className={'tlui-button__label' as any}>{msg(saveFileAction.label!)}</span>
+							{saveFileAction.kbd && <Kbd>{saveFileAction.kbd}</Kbd>}
+						</DropdownMenu.Item>
+						<DropdownMenu.Item
+							type="menu"
+							id="fork"
+							onClick={() => forkAction.onSelect('document-name')}
+						>
+							<span className={'tlui-button__label' as any}>{msg(forkAction.label!)}</span>
+						</DropdownMenu.Item>
+					</DropdownMenu.Group>
+				</DropdownMenu.Content>
+			</DropdownMenu.Root>
+		</div>
+	)
+})
+
+function DocumentTopZoneContainer({ children }: { children: ReactNode }) {
+	const ref = useRef<HTMLDivElement>(null)
+
+	const updateLayout = useCallback(() => {
+		const element = ref.current
+		if (!element) return
+
+		const layoutTop = element.parentElement!.parentElement!
+		const leftPanel = layoutTop.querySelector('.tlui-layout__top__left')! as HTMLElement
+		const rightPanel = layoutTop.querySelector('.tlui-layout__top__right')! as HTMLElement
+
+		const totalWidth = layoutTop.offsetWidth
+		const leftWidth = leftPanel.offsetWidth
+		const rightWidth = rightPanel.offsetWidth
+		// ignore the width of the button:
+		const selfWidth = element.offsetWidth - 44
```

Comment:
the 44, 12, and 16 numbers are a bit magical - can you name this as const's that might help explain what they're accounting for?